exercises/practice/sum-of-multiples/.approaches/filter-for-multiples/content.md

```diff
@@ -0,0 +1,233 @@
+
+```python
+is_even = lambda n: n % 2 == 0
+numbers = range(20)                      # 0, 1, …, 19
+even_numbers = filter(is_even, numbers)  # 0, 2, …, 18
+sum(numbers)       # ⟹ 190
+sum(numbers)       # ⟹ 190
+sum(even_numbers)  # ⟹ 90
+sum(even_numbers)  # ⟹ 0
+```
+
+Here, `sum` iterates over both `numbers` and `even_numbers` twice.
+
+In the case of `numbers` everything is fine.
+Even after looping through the whole of `numbers`, all its elements are still there, and so `sum` can ask to see them again without problem.
+
+The situation with `even_numbers` is less simple.
+To use the _cursor_ analogy: after going through all of `even_number`'s 'elements' &ndash; actually elements of `numbers` &ndash; the cursor has moved all the way to the right.
+It cannot move backwards, so if you wish to iterate over all even numbers again then you need a new cursor.
+We say that the `even_numbers` iterator is _exhausted_. When `sum` asks for its elements again, `even_numbers` comes up empty and so `sum` returns `0`.
+
+Had the highlighted solution not used `filter`, it might have looked like this:
+
+```python
+def sum_of_multiples(limit, factors):
+    is_multiple = lambda n: any(n % f == 0 for f in factors if f != 0)
+    multiples = [candidate for candidate in range(limit) if is_multiple(candidate)]
+    return sum(multiples)
+```
+
+This variant stores all the multiples in a `list` before summing them.
+Such a list can become very big.
+For example, if `limit = 1_000_000_000` and `factors = [1]` then `multiples` will take up 8 gigabytes of memory!
+It is to avoid unnecessarily creating such large intermediate data structures that iterators are often used.
+
+
+### A function expression: `lambda`
+
+Typically, when using higher-order functions like `filter` and `map`, the function to pass as an argument does not yet exist and needs to be defined first.
+
+The standard way of defining functions is through the `def` statement:
+
+```python
+def name(parameters):
+    statements
+```
+
+Downsides of this construct include
+
+- the syntax can be a bit bulky
+- it requires coming up with a fresh name
+
+These qualities can be quite bothersome when you just need a simple function of no particular significance for single use only.
+In situations like this you might like to use a **lambda expression** instead.
+
+A lambda expression is a specific kind of expression that evaluates to a function.
+It looks like this:
+
+```python
+lambda parameters: expression  # general form
+lambda a, b, x: a * x + b      # specific example
+```
+
+This latter lambda expression evaluates to a function that takes three arguments (`a`, `b`, `x`) and returns the value `a * x + b`.
+Except for not having a name, it is equivalent to the function defined by
+
+```python
+def some_name(a, b, x):
+    return a * x + b
+```
+
+A lambda expression need not necessarily be passed as an argument.
+It can also be applied to arguments immediately, or assigned to a variable:
```

Comment:
PEP 8 specifically sets out that you **should not assign a name to a lambda** (emphasis mine):

![image](https://user-images.githubusercontent.com/5923094/229324927-2c7b42d4-d9b5-4c57-8097-12a3107f012b.png)

Reply:
I agree that we should not encourage learners to do this. I will rework the text to avoid this.

The current text tries hard to explain that _lambda's are expressions_. However, I think this would become significantly harder to convey without the assignment example. There isn't really a lot one can do with a function. Immediately applying a lambda is unfamiliar, as is passing it as an argument (especially so if one is unfamiliar with passing named functions). Assignment on the other hand is _very_ familiar. If lambda's are only ever passed as argument, lambda's being expressions loses [productivity](https://en.wikipedia.org/wiki/Productivity_(linguistics)) (in the linguistic sense). Then they might as well be interpreted as a grammatical element that only occurs in argument lists. (You and I, knowing about lambda's being expressions, know to also expect them in `return`, `yield`, `… if … else …`, etc.)

Please let me know your thoughts about the latest commit. Note: the `lambda` and `filter` sections should probably swap places now (because of the `filter`-less example); I haven't done this yet to keep the diff simple. The changed code samples seem worse to me; I am tempted to rip out the `lambda` entirely.

Reply:
I came across [this DataQuest tutorial](https://www.dataquest.io/blog/tutorial-lambda-functions-in-python/#:~:text=A%20lambda%20function%20is%20an,and%20returns%20only%20one%20expression.&text=Note%20that%2C%20unlike%20a%20normal,a%20lambda%20function%20with%20parentheses.) that I think does a nice job, but even it assigns a variable to a `lambda` -- then says "don't do that!" , and encourages different practice.  One thing I though was really interesting was that they showed `if/else` in a `lambda` there.

So I know I am asking for something that might not be do-able in a teaching context.  But I think we should try, because all those `yacht` solutions assigning constants to `lambdas` are making me want to _scream_.  

I'll review your changes today, and have another think on it.  Wow, is that upcoming concept about `lambdas` going to be a b*$ch to write.....

Reply:
> [&hellip;] all those `yacht` solutions assigning constants to `lambdas` are making me want to _scream_.

Might this be catchable with the Analyzer?

> Wow, is that upcoming concept about `lambdas` going to be a b*$ch to write.....

I'm not so worried. There isn't that much more to them than already explained here, is there?

Reply:
> > […] all those `yacht` solutions assigning constants to `lambdas` are making me want to _scream_.
> 
> Might this be catchable with the Analyzer?

Oh, YESH.  Buuut...first I need to upgrade to Python 3.11.  Then I need to get the analyzer turned on for practice exercises.  Then I can enable that PyLint Rule for `Yacht`.  Maybe when I get there, I can ping you and we can see if there is messaging we can add that isn't just "PEP 8 says NO".  😁 


> > Wow, is that upcoming concept about `lambdas` going to be a b*$ch to write.....
> 
> I'm not so worried. There isn't that much more to them than already explained here, is there?

_You volunteering to get the [`lambda` concept and exercise](https://github.com/exercism/python/pull/2589) over the finish line?? (or start over with something else?)_ Because it can be ALL YOURS.  😄

Reply:
Just my two cents here. While I generally agree with PEP-8 and most the linter outputs ... I think `lambda` is a fairly concise and clean way to solve Yacht. There are exceptions to every rule and I don't think using `lambda` for Yacht should trigger a warning.

Reply:
@IsaacG - While I appreciate your two cents here, my comment about Yacht here was an _aside_.  If you'd like to debate a `lambda` check for Yacht, that's best done on the forum.